boards/arm/efm32hg_slstk3400a/efm32hg_slstk3400a.dts

```diff
@@ -57,5 +57,7 @@
 &usart1 {
 	current-speed = <115200>;
 	location = <4>;
+	pin-rx = <GPIO_GECKO_PORT_A 0>;
```

Comment:
1. I would prefer if we could merge the location index and the actual used pin into one property, something like:
`location-tx = <4 GPIO_GECKO_PORT_A PIN_0>`.
It would be clearer that these information belong together.

2. I don't know if it is a good idea: To ensure that only valid combinations are used, we could provide macros for all valid combinations in a soc header, like in Linux for the pins and the alternative functions on these pins. This would be a lot of defines, though (32 for USART-TX, 32 for USART-RX etc).

Reply:
I've replied to 1. in the main thread to give it more visibility. As for 2. for me a setup where user has to fill in all the pin configuration data himself is acceptable. Even if it would be user's responsibility to verify with the datasheet that the data are consistent. However, a solution that automates some tasks would definitely be better. I have an impression though it wouldn't be that easy to implement. Could you point me to the example from Linux header files you referred to?

Reply:
> Could you point me to the example from Linux header files you referred to?

Pin and function definition like in this file:
https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/imx6q-pinfunc.h#L709
These are used in the device tree to configure the pins for the actual board:
https://elixir.bootlin.com/linux/latest/source/arch/arm/boot/dts/imx6qdl-udoo.dtsi#L134

Reply:
I'm not opposed to the location-tx solution, but what would the DTS look like for proper pin mux definition?

Reply:
The standard approach would be to use [pinctrl-bindings](https://www.kernel.org/doc/Documentation/devicetree/bindings/pinctrl/pinctrl-bindings.txt). That's also what NXP and ST SoC are partly implementing. The DTS definitions are mostly there, however not used at the moment. E.g. [nxp_k6x.dtsi](https://github.com/zephyrproject-rtos/zephyr/blob/635191ac8c563990abb9d471aed0d97a4b86cc95/dts/arm/nxp/nxp_k6x.dtsi#L213)

In case of SiLabs SoCs the pinctrl approach is a bit tricky though since the location property must be defined in the driver, the matching GPIO pin in pinctrl. Keeping this in sync is not going to be easy. So the approach in this PR is actually easier to use, just not standard.

Reply:
> I'm not opposed to the location-tx solution, but what would the DTS look like for proper pin mux definition?

We could have something like:
`location-tx = <SOC_USART0_LOCATION_TX_5_PORT_A_PIN0>`
in the dts. This means for the peripheral USART0, we want to use the tx location 5. According to the datasheet tx location 5 is on GPIO-Port A, Pin 0.

The define `SOC_USART0_LOCATION_TX_5_PORT_A_PIN0` would be pulled from a header file for the specific SOC (or SOC family). We could even encode the correct pin configuration in this define, e.g. Pull-Ups, Slew-Rate, Push-Pull, Open-Collector etc.

Reply:
Can you guys mock up what the .dts's would look like for proper pin control, just keep it simple to a single u{s}art instance.  The one issue I have with the `location-tx = <4 GPIO_GECKO_PORT_A PIN_0>` proposal is that we end up duplicating SoC info in the board.dtsi.  Wondering if we can do anything about that or not.

Reply:
I've noticed the last message too late and pushed already the new version with `location-tx = <4 GPIO_GECKO_PORT_A PIN_0>` approach. I don't think we are going to duplicate SoC info as the board.dts file is the only one which will contain pin configuration data. But maybe I'm missing something.

I can prepare a more detailed mock up of the pinctrl approach but in short the definition of the DTS uart0 node in e.g. efr32fg1p.dtsi would have

```
	location = <0>;
	pinctrl-0 = <&uart0_default>;
```
and then in the pinctl node we would need to provide definition for `uart0_default` child node. How exactly the `uart0_default` should look like is not clear as the Zephyr pinmux framework is not finished yet.

The problem with the pinctrl approach is that modification of location parameter value has to be met with a matching change to the data in uart0_default node. On a typical SoC the routing is controlled fully by a pin controller/GPIO module by setting the so called pin 'function' parameter. On SiLabs SoCs 'function' register is called 'location' and it belongs to the register space of the controlling hardware module, UART, I2C, etc.

I'm not saying that pinctrl approach is impossible to implement but it's going to be tricky.

Reply:
> We could have something like:
> `location-tx = <SOC_USART0_LOCATION_TX_5_PORT_A_PIN0>`
> in the dts.

That looks interesting but I would need to think a bit more about it. I propose to do it in the next stage.

> We could even encode the correct pin configuration in this define, e.g. Pull-Ups, Slew-Rate, Push-Pull, Open-Collector etc.

We need to let user define custom pin configuration such as pull-up, slew-rate, open-collector, etc. as it depends on the board design. I thought about adding an optional 4-th parameter to the `location-` property. If the 4-th parameter is defined the driver will use it, if it's missing the driver will fall back to the default configuration. Also this feature I propose to add in a later stage.

Reply:
>>    We could even encode the correct pin configuration in this define, e.g. Pull-Ups, Slew-Rate, Push-Pull, Open-Collector etc.

> We need to let user define custom pin configuration such as pull-up, slew-rate, open-collector, etc. as it depends on the board design. I thought about adding an optional 4-th parameter to the location- property. If the 4-th parameter is defined the driver will use it, if it's missing the driver will fall back to the default configuration.

Yes, that is a good idea.

> Also this feature I propose to add in a later stage.

That is fine, i am all for incremental changes.